via/resources.py

```diff
@@ -44,25 +54,41 @@ def url_from_path(self):
 
         return self._normalise_url(url)
 
-    @classmethod
-    def _normalise_url(cls, url):
-        """Return a normalized URL from user input.
 
-        Take a URL from user input (eg. a URL input field or path parameter) and
-        convert it to a normalized form.
-        """
+class QueryURLResource(_URLResource):
+    """Resource for routes expecting urls from the query."""
+
+    def url_from_query(self):
+        """Get the 'url' parameter from the query.
 
+        :return: The URL as a string
+        :raise HTTPBadRequest: If the URL is missing or empty
+        :raise BadURL: If the URL is malformed
+        """
         try:
-            parsed = urlparse(url)
-        except ValueError as exc:
-            raise BadURL(url) from exc
+            url = self._request.params["url"].strip()
+        except KeyError as err:
+            raise HTTPBadRequest("Required parameter 'url' missing") from err
 
-        if not parsed.scheme:
-            if not parsed.netloc:
-                # Without a scheme urlparse often assumes the domain is the
-                # path. To prevent this add a fake scheme and try again
-                return cls._normalise_url("https://" + url)
+        if not url:
+            raise HTTPBadRequest("Required parameter 'url' is blank")
 
-            url = parsed._replace(scheme="https").geturl()
+        return self._normalise_url(url)
 
-        return Configuration.strip_from_url(url)
+
+def get_original_url(context):
+    """Get the original URL from a provided context object (if any)."""
+
+    if isinstance(context, QueryURLResource):
+        getter = context.url_from_query
+    elif isinstance(context, PathURLResource):
+        getter = context.url_from_path
+    else:
+        return None
+
+    try:
+        return getter()
+    except BadURL as err:
+        return err.url
+    except HTTPBadRequest:
+        return None
```

Comment:
We could try and facilitate this with a `dont_raise=True` or something in the methods, but we're unlikely to be able to write that more concisely than this.

Not sure if we should be catching `Exception` here instead though?

Reply:
Instead of `HTTPBadRequest`? I think we rather see them bubble up to sentry and check there if we find something interesting.

Reply:
We should report them anyway in the error handling so this should only be if our serialisation fails? I guess we'll see!